File: leetcode/python/0069_sqrt_x.py
class Solution(object):
    def mySqrt(self, x):
        """
        :type x: int
        :rtype: int
        """
        if x == 0: return 0
        if x == 1 or x == 2: return 1

        l = 1
        r = x
        middle = 0

        while (l <= r):
            middle = l + (r - l) / 2

            if int(x / middle) == middle:
                return middle
            elif x / middle < middle:
                r = middle - 1
            elif x / middle > middle:
                l = middle + 1
        
        return l - 1

File: leetcode/java/0069_sqrt_x.java
class Solution {
    public int mySqrt(int x) {
        if (x == 0) return 0;
        if (x == 1) return 1;

        int low = 0, high = x, middle = 0;

        while (low <= high) {
            middle = low + (high - low) / 2;

            if (x / middle == middle) return middle;
            else if (x / middle < middle) high = middle - 1;
            else if (x / middle > middle) low = middle + 1;
        }

        return low - 1;
    }
}